### ISSUE TITLE
[BUG] MySQL connection charset/collation hardcoded in DoctrineListener

### ISSUE THREAD

**Issue by maetthu** [bolt/bolt#6430]
As I need to save 4-byte utf8 values like emojis in content, I need to change the connection charset to utf8mb4. Though there is a possibility to change charset & collation in the database section of config.yml  (see #4485 and #4498), this does only apply for schema changes. The connection charset and collation is hardcoded  in [DoctrineListener](https://github.com/bolt/bolt/blob/6307ce400eb9e74b7f6140bc7a8e684ce38b50c4/src/EventListener/DoctrineListener.php#L82).

I don't know the correct/clean way to fix this directly in Bolt, but I'm currently using following workaround extension to overwrite the default DoctrineListener and allowing to set the connection charset & collation trough config.yml: https://gist.github.com/maetthu/ae4a13db20c5a0480c7d975361f2b2c6

**Comment by CarsonF** [comment]
We'd happily accept a PR for that. Only thing I would change is pass them into the class constructor (before logger).

**Comment by maetthu** [comment]
Sure, will do that then.

One problem: The config defaults use _utf8_unicode_ci_ for the collation option, though the `SET NAMES utf8` used so far implicitly sets the connection collation to the default for utf8, which is _utf8_general_ci_. If I set the connection collation to the config value, it would break backwards compatibility because the collate option was never used outside schema building and never was the collation used for the connection. If, OTOH, there is a check for the default value and make it behave like before in that case, it will not work when someone explicitly sets _utf8_unicode_ci_ in the config. Is there a way to detect, if a config value is actually set explicitly or just a default? Or maybe just introduce a new config option for the connection collation?

**Comment by GawainLynch** [comment]
> Is there a way to detect, if a config value is actually set explicitly or just a default? 

No is both the short and long answer there.

> Or maybe just introduce a new config option for the connection collation?

Please no. Those options are intended to match The Silex, Symfony & Doctrine combination, and I am in the middle of working though troubles on two we vary on already.

---

@maetthu I'm jet lagged as heck, so am having a _slight_ problem seeing where you're going with  of this … I may have also had my brain eaten by a Drop Bear recently :wink: 

Where you're invoking 
```php
$db->executeQuery(sprintf('SET NAMES %s COLLATE %s', $this->charset, $this->collation));
```

…then aren't you just needing to update your local db/tables:

```sql
ALTER DATABASE my_bolt_db CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
ALTER TABLE bolt_my_contenttype CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
```

**Comment by maetthu** [comment]
@GawainLynch sorry about the confusion :wink:

MySQL charsets can be pretty messy, even utf8 because it isn't proper utf8 :unamused:. The server, database, table and column charset & collation are just one part, sometimes they matter, sometimes they don't - if you force another charset on the connection using `SET NAMES` some conversion is made. But because `SET NAMES utf8` is hardcoded within bolt, and 4-byte utf8 values can't be converted to 3-byte utf8, saving content with 4-byte utf8 values (e.g. 😁 or 😂) fail with errors like this:

`SQLSTATE[HY000]: General error: 1366 Incorrect string value: '\xF0\x9F\x98\x81\xF0\x9F...' for column 'teaser' at row 1`

even though database, table and columns are all set to utf8mb4 and a matching collation.

```SQL
> show create database mtv\G
*************************** 1. row ***************************
       Database: mtv
Create Database: CREATE DATABASE `mtv` /*!40100 DEFAULT CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci */
1 row in set (0.00 sec)
> show create table bolt_articles\G
*************************** 1. row ***************************
       Table: bolt_articles
Create Table: CREATE TABLE `bolt_articles` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `slug` varchar(128) COLLATE utf8mb4_unicode_ci NOT NULL,
  `datecreated` datetime NOT NULL,
  `datechanged` datetime NOT NULL,
  `datepublish` datetime DEFAULT NULL,
  `datedepublish` datetime DEFAULT NULL,
  `username` varchar(32) COLLATE utf8mb4_unicode_ci DEFAULT '',
  `ownerid` int(11) DEFAULT NULL,
  `status` varchar(32) COLLATE utf8mb4_unicode_ci NOT NULL,
  `templatefields` longtext COLLATE utf8mb4_unicode_ci COMMENT '(DC2Type:json_array)',
  `headline` varchar(256) COLLATE utf8mb4_unicode_ci DEFAULT '',
  `subheadline` varchar(256) COLLATE utf8mb4_unicode_ci DEFAULT '',
  `facebookrepost` int(11) NOT NULL DEFAULT '0',
  `showteaser` longtext COLLATE utf8mb4_unicode_ci,
  `showpopular` tinyint(1) NOT NULL DEFAULT '0',
  `showads` tinyint(1) NOT NULL DEFAULT '0',
  `image` longtext COLLATE utf8mb4_unicode_ci COMMENT '(DC2Type:json_array)',
  `teasersmall` longtext COLLATE utf8mb4_unicode_ci COMMENT '(DC2Type:json_array)',
  `teaser` longtext COLLATE utf8mb4_unicode_ci,
  `lead` longtext COLLATE utf8mb4_unicode_ci,
  `legacyid` longtext COLLATE utf8mb4_unicode_ci,
  `legacyurl` longtext COLLATE utf8mb4_unicode_ci,
  PRIMARY KEY (`id`),
  KEY `IDX_E9659B55989D9B62` (`slug`),
  KEY `IDX_E9659B55AFBA6FD8` (`datecreated`),
  KEY `IDX_E9659B55BE74E59A` (`datechanged`),
  KEY `IDX_E9659B55A5131421` (`datepublish`),
  KEY `IDX_E9659B55B7805520` (`datedepublish`),
  KEY `IDX_E9659B557B00651C` (`status`),
  KEY `IDX_E9659B55D8E1B211` (`showpopular`)
) ENGINE=InnoDB AUTO_INCREMENT=5933 DEFAULT CHARSET=utf8mb4 COLLATE=utf8mb4_unicode_ci ROW_FORMAT=DYNAMIC
1 row in set (0.00 sec)
```

So what I'm trying to do is being able to change the hardcoded `SET NAMES utf8` to a different charset, utf8mb4 in my case. Saving 4-byte utf8 content works when I override the bolt init SQL to this one:

```SQL
SET NAMES utf8mb4 COLLATE utf8mb4_unicode_ci 
```

**Comment by GawainLynch** [comment]
Side note: Reading [`\Doctrine\DBAL\Platforms\MySqlPlatform::buildTableOptions()`](https://github.com/doctrine/dbal/blob/2.5/lib/Doctrine/DBAL/Platforms/MySqlPlatform.php#L472-L519) on a Friday morning, thrilling reading … not even 08:00 and I think I need :beer: :wink: 

Our use of `$db->executeQuery('SET NAMES utf8');` & `$db->executeQuery('SET CHARACTER_SET_CONNECTION = utf8);` seems like a historical workaround, and probably(?) should not be used, nor should you need your override, as `buildTableOptions()` set all this as passed in to the connection options when you set:
```
database:
    driver: mysql
    databasename: bolt_testing
    username: bolt_testing
    password: bolt_testing
    charset:  utf8mb4
    collate:  utf8mb4_unicode_ci
```

Now here's the catch … to make it all work, the database needs to be first created as `utf8mb4`, on the command line … using the web UI will running to a redirect loop currently as there is a check for the existence of `bolt_users` … failure of said check redirects to first user … which is where we already are :koala: 

```
$ ./app/nut database:update
Modifications made to the database:
 - Created table `bolt_authtoken`. 
 - Created table `bolt_blocks`. 
 - Created table `bolt_cron`. 
 - Created table `bolt_entries`. 
 - Created table `bolt_field_value`. 
 - Created table `bolt_log_change`. 
 - Created table `bolt_log_system`. 
 - Created table `bolt_pages`. 
 - Created table `bolt_relations`. 
 - Created table `bolt_showcases`. 
 - Created table `bolt_taxonomy`. 
 - Table `bolt_users` is not present. 
Your database is now up to date.
```

**NOTE:** ` - Table `bolt_users` is not present.`

The problem being `bolt_users` creation fails with `Error in query (1071): Specified key was too long; max key length is 767 bytes` … (btw, dropping the unique index on emails — not an option — can work around that, but I digress)

Commenting out `charset:` and `collate:`, then re-running `$ ./app/nut database:update` will build the user's table (as UTF8), and I am able to save `𠜎  𢫕 𢭃 𢯊 𢱑` and have them render perfectly in the front-end …

So I can see the value in dropping the hard coding (thoughts @rossriley @CarsonF ?), as Doctrine already provides them as UTF8 if not passed in via YAML options … but … 

> Hello Support Hell, 
> All the things are broken in Bolt.
> Regards, 
> Your friendly `utf8mb4` user.

**Comment by maetthu** [comment]
@GawainLynch I don't think this is the first (and last) time someone had the urge to drink :beers: (or something stronger) after dealing with charset issues in MySQL - been there :smirk:

> Our use of $db->executeQuery('SET NAMES utf8'); & $db->executeQuery('SET CHARACTER_SET_CONNECTION = utf8); seems like a historical workaround, and probably(?) should not be used, nor should you need your override, as buildTableOptions() set all this as passed in to the connection options when you set:

`$db->executeQuery('SET CHARACTER_SET_CONNECTION = utf8);` is definitely not needed anymore, because `SET NAMES` is the same as setting _character_set_client_, _character_set_results_, _character_set_connection_ individually, thus this is redundant. I'm not sure you could drop `SET NAMES` though - unless the server and client configuration are already properly set to utf8 (which lots are not, latin1 is still the default for many installations), you need to tell MySQL that everything PHP sends is utf8.

> The problem being bolt_users creation fails with `Error in query (1071): Specified key was too long; max key length is 767 bytes` … (btw, dropping the unique index on emails — not an option — can work around that, but I digress)

There are different solutions, the easiest (and probably also the safest in terms of support for different environments) would be to reduce field length to <= 191 (I think ~100 would be more than enough for an email address) or, what I did for my installation, set the [innodb_large_prefix](https://dev.mysql.com/doc/refman/5.7/en/innodb-parameters.html#sysvar_innodb_large_prefix) option in MySQL config and do an `ALTER TABLE `table`  ROW_FORMAT=DYNAMIC;` for all tables before converting everything to utf8mb4 or directly force the row format when creating the table. But that's probably not an option for most users in hosted environments. 

I see you point about it being a support hell though. OTOH, not supporting utf8mb4 is not really ideal as well. I had to fix it because my users embed a lot of Instagram posts and apparently you have to include at least a dozen emojis in your post to be a proper Instagram user :unamused:. Maybe we are not a very typical use case, but I think some other users may run into this problem as well. I could release my charset overwriting extension if that helps.

**Comment by GawainLynch** [comment]
>> The problem being bolt_users creation fails with Error in query (1071): Specified key was too long; max key length is 767 bytes … (btw, dropping the unique index on emails — not an option — can work around that, but I digress)

> … would be to reduce field length to <= 191 (I think ~100 would be more than enough for an email address)

I am very much afraid not, the [spec requires 254 characters](https://tools.ietf.org/html/rfc3696#section-4.3) and we've been bitten before believe it or not. Also both user name, and email address, must individually be unique.

---

> you need to tell MySQL that everything PHP sends is utf8.

That should be what DBAL is setting up with it's connection parameters, it was when I was running xdebug over it, but didn't test against a `latin1 ` base though.

---

> and apparently you have to include at least a dozen emojis in your post to be a proper Instagram user

Quote of the day!

--- 

Personally I have no objections to making the connection parameters configurable as we've discussed. But at least for now, we'd probably have to consider `utf8mb4` "experimental" and most certainly "expert" category (i.e. "If it breaks, you get to keep both pieces").

**Comment by maetthu** [comment]
Sorry about my late response :blush: 

>> … would be to reduce field length to <= 191 (I think ~100 would be more than enough for an email address)

> I am very much afraid not, the spec requires 254 characters and we've been bitten before believe it or not. Also both user name, and email address, must individually be unique.

I see... unlike IPv4 where address space will "never" run out, they specified something back in those days which will never run out of space for real :grin: 

Converting everything to utf8mb4 needs some manual labor then:

```SQL
ALTER DATABASE databasename CHARACTER SET = utf8mb4 COLLATE = utf8mb4_unicode_ci;
-- for each table
ALTER TABLE tablename ROW_FORMAT=DYNAMIC;
ALTER TABLE tablename CONVERT TO CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci;
```
and setting the innodb_large_prefix option in the server config. 

>> you need to tell MySQL that everything PHP sends is utf8.

> That should be what DBAL is setting up with it's connection parameters, it was when I was running xdebug over it, but didn't test against a latin1 base though.

You're right - it's working fine for me as well w/o manually sending a SET NAMES in the connection listener. 

> Personally I have no objections to making the connection parameters configurable as we've discussed. But at least for now, we'd probably have to consider utf8mb4 "experimental" and most certainly "expert" category (i.e. "If it breaks, you get to keep both pieces").

Not having to overwrite the postConnect listener to support another charset than utf8 would be nice, yes. Or, as it does involve some manual changes to the database, just list the necessary steps in the docs.

**Comment by GawainLynch** [comment]
At this stage I have submitted a PR, that if ever merged (can take **hours** around here!!! :wink:), will need follow up docs.

I don't suppose you feel like adding a HOWTO on `utf8mb4` in the [HOWTO section](https://github.com/bolt/docs/tree/release/3.0/docs/howto) of the docs? Waiting on my schedule has been known to cost years of human time :laughing:

**Comment by maetthu** [comment]
Yeah, sure, I can do that! :smiley:

**Comment by bobdenotter** [comment]
Fixed by #6533
